pkg/controller/pxc/users.go

```diff
@@ -38,159 +39,312 @@ type ReconcileUsersResult struct {
+	//update internalSecrets
+	internalSecrets.Data[users.UserRoot] = secrets.Data[users.UserRoot]
+	err = r.client.Update(context.TODO(), internalSecrets)
+	if err != nil {
+		return errors.Wrap(err, "update internal users secrets operator user password")
+	}
+
+	upRes.restartPXC = true
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) manageOperatorAdminUser(cr *api.PerconaXtraDBCluster, secrets, internalSecrets *corev1.Secret) error {
+	pass, existInSys := secrets.Data[users.UserOperator]
+	_, existInInternal := internalSecrets.Data[users.UserOperator]
+	if existInSys && !existInInternal {
+		if internalSecrets.Data == nil {
+			internalSecrets.Data = make(map[string][]byte)
 		}
+		internalSecrets.Data[users.UserOperator] = pass
+		return nil
+	}
+	if existInSys {
+		return nil
 	}
 
-	restarts, err := r.manageSysUsers(cr, &sysUsersSecretObj, &internalSysSecretObj)
+	pass, err := generatePass()
 	if err != nil {
-		return nil, errors.Wrap(err, "manage sys users")
+		return errors.Wrap(err, "generate password")
 	}
+	addr := cr.Name + "-pxc." + cr.Namespace
+	if cr.CompareVersionWith("1.6.0") >= 0 {
+		addr = cr.Name + "-pxc-unready." + cr.Namespace + ":33062"
+	}
+	um, err := users.NewManager(addr, users.UserRoot, string(secrets.Data[users.UserRoot]), cr.Spec.PXC.ReadinessProbes.TimeoutSeconds)
+	if err != nil {
+		return errors.Wrap(err, "new users manager")
+	}
+	defer um.Close()
 
-	internalSysSecretObj.Data = sysUsersSecretObj.Data
-	err = r.client.Update(context.TODO(), &internalSysSecretObj)
+	err = um.CreateOperatorUser(string(pass))
 	if err != nil {
-		return nil, errors.Wrap(err, "update internal sys users secret")
+		return errors.Wrap(err, "create operator user")
 	}
 
-	result := &ReconcileUsersResult{
-		updateReplicationPassword: restarts.updateReplicationPass,
+	secrets.Data[users.UserOperator] = pass
+	internalSecrets.Data[users.UserOperator] = pass
+
+	err = r.client.Update(context.TODO(), secrets)
+	if err != nil {
+		return errors.Wrap(err, "update sys users secret")
+	}
+	err = r.client.Update(context.TODO(), internalSecrets)
+	if err != nil {
+		return errors.Wrap(err, "update internal users secret")
 	}
 
-	if restarts.restartProxy {
-		result.proxysqlAnnotations = map[string]string{"last-applied-secret": newSecretDataHash}
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) handleMonitorUser(cr *api.PerconaXtraDBCluster, secrets, internalSecrets *corev1.Secret, upRes *userUpdateActions) error {
+	user := &users.SysUser{
+		Name:  users.UserMonitor,
+		Pass:  string(secrets.Data[users.UserMonitor]),
+		Hosts: []string{"%"},
 	}
-	if restarts.restartPXC {
-		result.pxcAnnotations = map[string]string{"last-applied-secret": newSecretDataHash}
+
+	//update grants
+	if cr.CompareVersionWith("1.6.0") >= 0 {
+		// monitor user need more grants for work in version more then 1.6.0
+		err := r.updateMonitorUserGrant(cr, internalSecrets)
```

Comment:
with this implementation, we won't update required grants for these users. we were running these functions even if secret data is not changed.

Reply:
Why do we need to upgrade them even if nothing changed?

Reply:
because these functions are there to fix grants on deployments with older versions. we can make this one the default by updating grants in entrypoint but for example for `replication` user we still need to do this. this is not bound to a password change.

Reply:
Aha, got it.

Reply:
Done